src/Algebra/Lattice/Stacked.hs

```diff
@@ -0,0 +1,100 @@
+{-# LANGUAGE DeriveFunctor       #-}
+{-# LANGUAGE DeriveGeneric       #-}
+{-# LANGUAGE DeriveTraversable   #-}
+{-# LANGUAGE FlexibleContexts    #-}
+{-# LANGUAGE Safe                #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeOperators       #-}
+module Algebra.Lattice.Stacked (
+    Stacked(..)
+  , leftOnTopOfRight
+  , rightOnTopOfLeft
+  ) where
+
+import Prelude ()
+import Prelude.Compat
+
+import Algebra.Lattice
+import Algebra.PartialOrd
+
+import Control.DeepSeq       (NFData (..))
+import Control.Monad         (ap)
+import Data.Data             (Data, Typeable)
+import Data.Hashable         (Hashable (..))
+import Data.Universe.Class   (Finite (..), Universe (..))
+import GHC.Generics          (Generic, Generic1)
+
+import qualified Test.QuickCheck as QC
+
+--
+-- Stacked
+--
+
+-- | Stacked two lattices, one on top of another. All minimal elements of upper lattice cover all maximal elements of lower lattice.
+data Stacked a b = Lower a
+              | Upper b
+    deriving ( Eq, Ord, Show, Read, Data, Typeable, Generic, Functor, Foldable, Traversable
+            , Generic1
+            )
+
+-- | Converts @Either@ to @Stacked@ assuming @Either@'s left on top of right
+leftOnTopOfRight :: Either l r -> Stacked r l 
+leftOnTopOfRight (Left l) = Upper l
+leftOnTopOfRight (Right r) = Lower r
+
+-- | Converts @Either@ to @Stacked@ assuming @Either@'s right on top of left
+rightOnTopOfLeft :: Either l r -> Stacked l r 
+rightOnTopOfLeft (Left l) = Lower l
+rightOnTopOfLeft (Right r) = Upper r
+
+instance Applicative (Stacked a) where
+    pure = Upper
+    (<*>) = ap
+
+instance Monad (Stacked a) where
+    Lower x >>= _ = Lower x
+    Upper x >>= f = f x
+
+instance (NFData a, NFData b) => NFData (Stacked a b) where
+    rnf (Upper x) = rnf x
+    rnf (Lower x) = rnf x
+
+instance (Hashable a, Hashable b) => Hashable (Stacked a b)
+
+instance (PartialOrd a, PartialOrd b) => PartialOrd (Stacked a b) where
+    leq (Upper x) (Upper y) = leq x y
+    leq (Upper _) _ = False
+    leq _ (Upper _) = True
+    leq (Lower x) (Lower y) = leq x y
+    comparable (Upper x) (Upper y) = comparable x y
+    comparable (Upper _) _ = True
+    comparable _ (Upper _) = True
+    comparable (Lower x) (Lower y) = comparable x y
+
+instance (Lattice a, Lattice b) => Lattice (Stacked a b) where
+    Upper x \/ Upper y = Upper (x \/ y)
+    u@(Upper _) \/ _ = u
+    _ \/ u@(Upper _) = u
+    Lower x \/ Lower y = Lower (x \/ y)
+    Lower x /\ Lower y = Lower (x /\ y)
+    l@(Lower _) /\ _ = l
+    _ /\ l@(Lower _) = l
+    Upper x /\ Upper y = Upper (x /\ y)
+
+instance (BoundedJoinSemiLattice a, Lattice b) => BoundedJoinSemiLattice (Stacked a b) where
+    bottom = Lower bottom
+
+instance (Lattice a, BoundedMeetSemiLattice b) => BoundedMeetSemiLattice (Stacked a b) where
+    top = Upper top
+
+instance (Universe a, Universe b) => Universe (Stacked a b) where
+    universe = (Lower <$> universe) ++ (Upper <$> universe)
```

Comment:
use [`+++`](https://hackage.haskell.org/package/universe-base-1.1.1/docs/Data-Universe-Helpers.html#v:-43--43--43-) to interleave

Maybe move this definition to `FInite` (using `universeF`) as there the `++` is ok, and the order might be useful for someone.

Reply:
Done as suggested.